<a href="https://colab.research.google.com/github/Byeon-MJ/TIL/blob/main/%5BPython%5D%20List_Delete_Time_Comparison.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python 리스트 삭제 속도 비교
리스트에서 특정 위치의 요소를 삭제할때 사용하는 대표적인 네가지 기능인 **slicing, pop, del, remove**의 처리 속도를 비교해보았다.

속도를 비교하게 된 계기는 프로그래머스에서 문제를 풀다가 효율성 테스트에서 계속 막혀서 성능 개선을 위해서였다. 

결론부터 말하자면 이런 내용을 몰랐을때 나는 Slicing 방법으로 리스트를 처리했었는데, del을 사용하니 속도가 많이 개선되었고 효율성 테스트를 통과할 수 있었다!!

성능을 비교하는 것이 주 목적이기에 각각의 사용법에 대한 것들은 생략하도록 하겠다.

* 아래는 해당 문제의 링크와 내 풀이를 포스팅한 내용이다. 어떤 문제인지와 풀이법이 궁금하다면 참고!

https://school.programmers.co.kr/learn/courses/30/lessons/133502

https://redmooncode.tistory.com/115

## Pop
길이가 500만인 리스트를 사용하였고, 시간 측정을 위해 time 모듈을 이용하였다.

아래 코드는 네가지 기능들 모두에 공통적으로 사용할 것이다.

In [55]:
import time

str_list = [i for i in range(5000000)]

먼저 pop의 성능을 알아보자. pop은 리스트에서 인덱스를 이용하여 값을 지운다. 그리고 지워진 인덱스의 값을 반환한다. del과는 다르게 지워진 값을 반환하기때문에 미세한 속도 차이가 생긴다고 한다.

리스트에서 값을 삭제한 후 리스트를 재조정한다.

In [53]:
# pop
start_time = time.time()
list_pop = str_list.copy()
list_pop.pop(0)
end_time = time.time()
print(f'pop : {end_time - start_time:.4f}초')

pop : 0.0450초


## del

del은 pop과 마찬가지로 인덱스를 이용하여 리스트의 값을 삭제한다.

리스트의 슬라이싱을 활용하여 삭제도 가능하며, pop과 다르가 리턴값이 없어서 미세하게 더 빠르다. 마찬가지로, 삭제한 후 리스트를 재조정한다.

In [51]:
# del
start_time = time.time()
list_del = str_list.copy()
del list_del[0]
end_time = time.time()
print(f'del : {end_time - start_time:.4f}초')

del : 0.0319초


## remove

remove는 위의 두개와는 달리 인덱스 입력으로 삭제하는 방식이 아니라 지우고자 하는 값을 입력하는 방법이다.  해당하는 값이 복수로 존재한다면 앞서 존재하는 것을 지운다. 

In [50]:
# remove
start_time = time.time()
list_remove = str_list.copy()
list_remove.remove(0)
end_time = time.time()
print(f'remove : {end_time - start_time:.4f}초')

remove : 0.0335초


## slicing

slicing은 엄밀히 말하면 리스트의 요소 삭제라기 보다는 입력한 범위로 다시 출력하는 방법이다. 

속도상 가장 느리다.

In [52]:
# slice
start_time = time.time()
list_slice = str_list.copy()
list_slice = list_slice[1:]
end_time = time.time()
print(f'slice : {end_time - start_time:.4f}초')

slice : 0.0909초


## 요약

- pop : 0.0450초
- del : 0.0319초
- remove : 0.0335초
- slicing : 0.0909초

`삭제` 를 기준으로는 del이 가장 빠르고 slicing이 가장 느리다

리스트 원본을 보존해야한다면 slicing 사용을 고려해도 괜찮다.

## Reference

https://brownbears.tistory.com/452